wzprof.go

```diff
@@ -10,30 +11,156 @@ import (
 	"unsafe"
 
 	"github.com/google/pprof/profile"
+	"github.com/tetratelabs/wazero"
 	"github.com/tetratelabs/wazero/api"
 	"github.com/tetratelabs/wazero/experimental"
 	"golang.org/x/exp/slices"
 )
 
+// Profiling mechanism for a given WASM binary. Entry point to generate
+// Profilers.
+type Profiling struct {
+	wasm []byte
+
+	filteredFunctions map[string]bool
+	symbols           symbolizer
+	stackIterator     func(mod api.Module, def api.FunctionDefinition, wasmsi experimental.StackIterator) experimental.StackIterator
+
+	isGo bool
+}
+
+// ProfilingFor a given wasm binary. The resulting Profiling needs to be
+// prepared after Wazero module compilation.
+func ProfilingFor(wasm []byte) *Profiling {
+	r := &Profiling{
+		wasm:    wasm,
+		symbols: noopsymbolizer{},
+		stackIterator: func(mod api.Module, def api.FunctionDefinition, wasmsi experimental.StackIterator) experimental.StackIterator {
+			return wasmsi
+		},
+	}
+
+	if binCompiledByGo(wasm) {
+		r.isGo = true
+		// Those functions are special. They use a different calling
+		// convention. Their call sites do not update the stack pointer,
+		// which makes it impossible to correctly walk the stack.
+		//
+		// https://github.com/golang/go/blob/7ad92e95b56019083824492fbec5bb07926d8ebd/src/cmd/internal/obj/wasm/wasmobj.go#LL907C18-L930C2
+		r.filteredFunctions = map[string]bool{
+			"_rt0_wasm_js":            true,
```

Comment:
A common model to avoid the boolean value
```suggestion
		r.filteredFunctions = map[string]struct{}{
			"_rt0_wasm_js":            {},
```
Then use it with the two-value lookup syntax:
```
_, exist := r.filteredFunctions[name]
```